### ISSUE TITLE
new Visual Studio v16.10 breaks existing spdlog library

### ISSUE THREAD

**Issue by gamagan** [gabime/spdlog#1955]
I have spdlog version 1.8.1

I made the mistake of upgrading from Visual Studio 16.9 to 16.10, and it broke my code. I get errors from spdlog library.

> spdlog/fmt/bundled/format.h(3510,29): error C2668: 'fmt::v7::make_format_args': ambiguous call to overloaded function
>
> spdlog/fmt/bundled/core.h(1424,43): message : could be 'fmt::v7::format_arg_store<context,MyType,_Ty> fmt::v7::make_format_args<context,MyType,_Ty>(const MyType &,const _Ty &)
>
> or       'auto std::make_format_args<context,MyType,_Ty>(const MyType &,const _Ty &)

Any idea how I can get around this issue?

**Comment by tt4g** [comment]
Was the error reported when you built your project?
From the error message it is not possible to distinguish if the problem is caused by your project code calling fmt through spdlog, or if there is a problem with the spdlog project itself.

**Comment by tt4g** [comment]
It seems that already fixed fmtlib/fmt#2295.
Should you external fmt library to solve this problem.

**Comment by damienhocking** [comment]
<format> is now part of VS 16.10

**Comment by janclod** [comment]
Hi, I am quite newbie so please take this into account while reading my comment.

I am experiencing and suffering form this issue on my own project. As far as I understand the closest solution is to include the latest fix (https://github.com/fmtlib/fmt/issues/2295) of the fmt library into spdlog. Is this correct? Do you have any other suggestions?

**Comment by tt4g** [comment]
@janclod Please replace the bundled fmt yourself.


Another way is to build fmt as an external library, and set the CMake variable `SPDLOG_FMT_EXTERNAL` or `SPDLOG_FMT_EXTERNAL_HQ` to `ON` when you build spdlog with cmake.

https://github.com/gabime/spdlog/blob/100f30043f33277122e0991c83845a2617172ffd/CMakeLists.txt#L176-L196

**Comment by janclod** [comment]
Hi tt4g, thanks for your reply. I believe that I tried something like this here: https://github.com/janclod/spdlog/tree/v1.8.1-alfa.

Is this what you mean? Simply copy-pasting the files from the fmt project into the bundled folder?

**Comment by tt4g** [comment]
@janclod Right, you need to replace the fmt source with the latest version.

**Comment by janclod** [comment]
@tt4g thanks, I will give it a try :)

**Comment by janclod** [comment]
@tt4g I do get an error when building

```
'void fmt::v7::format_system_error(fmt::v7::detail::buffer<char> &,int,const char *) noexcept': cannot convert argument 3 from 'const std::string' to 'const char *'	spdlog	C:\Users\ClaudioTiecher\source\repos\Version7\image_saver\external\xb-deps\xb-deps\janclod\spdlog\v1.8.1-alfa\include\spdlog\common-inl.h	59	
```

**Comment by tt4g** [comment]
@janclod This is known issue, fmt API has been changed in the latest release.
Look this comment: https://github.com/gabime/spdlog/issues/1941#issuecomment-841530108

And you can fix it with the following changes.

`include/spdlog/common-inl.h` line 56:

```diff
SPDLOG_INLINE spdlog_ex::spdlog_ex(const std::string &msg, int last_errno)
{
    memory_buf_t outbuf;
-    fmt::format_system_error(outbuf, last_errno, msg);
+    fmt::format_system_error(outbuf, last_errno, msg.c_str());
    msg_ = fmt::to_string(outbuf);
}
```

**Comment by janclod** [comment]
@tt4g thanks again for pointing me in the right direction, I am still getting more errors, I will first try to go through them myself and come back here if I need help.

**Comment by janclod** [comment]
@tt4g I am totally lost due to my lack of knowledge in C++, would you be able to help further?

```
'basic_data': looks like a function definition, but there is no parameter list; skipping apparent body	spdlog	C:\Users\ClaudioTiecher\source\repos\Version7\image_saver\external\xb-deps\xb-deps\janclod\spdlog\v1.8.1-alfa\src\fmt.cpp	61	
```

**Comment by tt4g** [comment]
@janclod `spdlog/src/fmt.cpp` has also been copied from the fmt library. Since this file doesn't seem to be updated in the your forked repository, you need to copy `fmt.cpp` from fmt repository.

**Comment by janclod** [comment]
@tt4g I am looking for `fmt.cpp` in the current fmt rep, but it is not there any longer.

**Comment by janclod** [comment]
@tt4g it looks like the `fmt.cpp` file is not there any longer. Any suggestion?

**Comment by tt4g** [comment]
@janclod Sorry, I checked the fmt repository and it seems that `src/format.cc` is renamed to `spdlog/src/fmt.cpp`.
It was written at the top of `spdlog/src/fmt.cpp`:
https://github.com/gabime/spdlog/blob/v1.8.5/src/fmt.cpp#L1

After copying `src/format.cc`, don't forget to add the spdlog macros at the beginning and end.

Example:

```cpp
// Copyright  here!

#ifndef SPDLOG_COMPILED_LIB
#error Please define SPDLOG_COMPILED_LIB to compile this file.
#endif

#if !defined(SPDLOG_FMT_EXTERNAL)
#include <spdlog/fmt/bundled/format-inl.h>

// `src/format.cc` contents here!

#endif // !SPDLOG_FMT_EXTERNAL
```

**Comment by janclod** [comment]
@tt4g no worries, you have been very helpful :)

I manage to build successfully! :D

**Comment by janclod** [comment]
@tt4g if I can contribute in any way, please let me know. I thought that it could be helpful to have my fork available for people facing this issue, maybe as a sort of temporary fix?

**Comment by tt4g** [comment]
@janclod When the next version of fmt is released, you can create a pull request (close #1941) to replace the fmt that spdlog bundles with the latest version.

Unfortunately, fmt has not been released since version 7.1.3 (which is the version that spdlog is bundled with), so you won't be able to create a pull request for a while.
However, don't worry about it, someone will work on the upgrade.
